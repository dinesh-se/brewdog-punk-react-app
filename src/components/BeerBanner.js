import React from 'react';

export default class BeerBanner extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      isLoaded: false,
      beerInfo: ''
    };
    this.getAnotherBeer = this.getAnotherBeer.bind(this);
  }

  fetchRandomBeer() {
    return fetch("https://api.punkapi.com/v2/beers/random")
      .then(res => res.json());
  }

  componentDidMount() {
    this.getAnotherBeer();
  }

  getAnotherBeer() {
    this.fetchRandomBeer()
      .then((response) => {
        this.setState(() => ({
          isLoaded: true,
          beerInfo: response[0]  
        }))
        console.log(this.state);
      }, (error) => {
        this.setState(() => ({
          isLoaded: true,
          error
        }));
      });
  }

  render() {
    const { error, isLoaded, beerInfo } = this.state;
    if (error) {
      return <div>Error: {error.message}</div>;
    } else if (!isLoaded) {
      return <div>Loading...</div>
    } else {
      return (
        <div>
          <img src={beerInfo.image_url} />
          <h3>{beerInfo.name}</h3>
          <p>{beerInfo.description}</p>
          <button onClick={this.getAnotherBeer}>Another Beer</button>
        </div>
      )
    }
  }
}