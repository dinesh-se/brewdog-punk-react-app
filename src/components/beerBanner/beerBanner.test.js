import React from 'react';
import { shallow } from 'enzyme';
import BeerBanner from './beerBanner';
import "isomorphic-fetch";

test('should render loader correctly', () => {
  const wrapper = shallow(<BeerBanner />);

  // default state - loading
  wrapper.setState({
    error: null,
    isLoaded: false,
    beerInfo: {}
  });

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('.banner-loading').text()).toBe('Loading...');
  expect(wrapper.find('.error-message')).toHaveLength(0);
  expect(wrapper.find('.banner-container')).toHaveLength(0);
});

test('should render error message correctly', () => {
  const wrapper = shallow(<BeerBanner />);

  // error state
  wrapper.setState({
    error: {
      message: "Error fetching data."
    },
    isLoaded: true,
    beerInfo: {}
  });

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('.error-message').text()).toContain('Error:');
  expect(wrapper.find('.banner-loading')).toHaveLength(0);
  expect(wrapper.find('.banner-container')).toHaveLength(0);
});

test('should render beer banner correctly', () => {
  const wrapper = shallow(<BeerBanner />);

  // success state
  wrapper.setState({
    error: null,
    isLoaded: true,
    beerInfo: {
      name: 'Beer 1',
      description: 'Beer description'
    }
  });

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('.beer-name').text()).toContain('Beer 1');
  expect(wrapper.find('.beer-desc').text()).toContain('Beer description');
  expect(wrapper.find('.banner-loading')).toHaveLength(0);
  expect(wrapper.find('.error-message')).toHaveLength(0);
});

test('should call another beer method on clicking button', () => {
  const wrapper = shallow(<BeerBanner />);
  wrapper.instance().getAnotherBeer = jest.fn();

  // success state
  wrapper.setState({
    error: null,
    isLoaded: true,
    beerInfo: {
      name: 'Beer 1',
      description: 'Beer description'
    }
  });

  wrapper.find('.button-alcoholic-beer').prop('onClick')();
  expect(wrapper.instance().getAnotherBeer).toHaveBeenCalled();
});

test('should fetch random beer successfully', (done) => {
  const wrapper = shallow(<BeerBanner />);
  const mockSucessResponse = [{
    name: 'Beer',
    description: 'Beer description',
    image_url: 'image.png'
  }];
  const mockJsonPromise = Promise.resolve(mockSucessResponse);

  wrapper.instance().fetchRandomBeer = jest.fn(() => mockJsonPromise);
  wrapper.instance().getAnotherBeer();

  expect(wrapper.instance().fetchRandomBeer).toHaveBeenCalled();
  wrapper.instance().fetchRandomBeer()
    .then(() => {
      expect(wrapper.state()).toEqual({
        isLoaded: true,
        beerInfo: mockSucessResponse[0],
        error: null
      });
      done();
    });
});

test('should set error message when API fails', (done) => {
  const wrapper = shallow(<BeerBanner />);
  const mockFailureResponse = {
    message: "Error"
  };
  const mockJsonPromise = Promise.reject(mockFailureResponse);

  wrapper.instance().fetchRandomBeer = jest.fn(() => mockJsonPromise);
  wrapper.instance().getAnotherBeer();

  expect(wrapper.instance().fetchRandomBeer).toHaveBeenCalled();
  wrapper.instance().fetchRandomBeer()
    .catch(() => {
      expect(wrapper.state()).toEqual({
        isLoaded: true,
        beerInfo: {},
        error: mockFailureResponse
      });
      done();
    });
});