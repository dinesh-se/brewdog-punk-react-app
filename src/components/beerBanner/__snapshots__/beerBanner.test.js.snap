// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render beer banner correctly 1`] = `
<div
  className="banner-container"
>
  <h3
    className="beer-name"
  >
    Beer 1
  </h3>
  <div
    className="flex-container"
  >
    <div
      className="image-container"
    >
      <img
        className="beer-image"
      />
    </div>
    <div
      className="beer-desc"
    >
      Beer description
    </div>
    <div
      className="switch-beer"
    >
      <button
        className="button primary-btn button-alcoholic-beer"
        onClick={[Function]}
      >
        Another Beer
      </button>
      <button
        className="button primary-btn"
        onClick={[Function]}
      >
        Random Non-Alcoholic Beer
      </button>
    </div>
  </div>
</div>
`;

exports[`should render error message correctly 1`] = `
<div
  className="error-message"
>
  Error: 
  Error fetching data.
</div>
`;

exports[`should render loader correctly 1`] = `
<div
  className="banner-loading"
>
  Loading...
</div>
`;
